pkg/controllers/provisioningv2/harvestercleanup/controller.go

```diff
@@ -0,0 +1,157 @@
+package harvestercleanup
+
+import (
+	"context"
+
+	"github.com/rancher/wrangler/v3/pkg/data"
+	corecontrollers "github.com/rancher/wrangler/v3/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	k8dynamic "k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/rancher/rancher/pkg/capr"
+	"github.com/rancher/rancher/pkg/controllers/capr/machineprovision"
+	capicontrollers "github.com/rancher/rancher/pkg/generated/controllers/cluster.x-k8s.io/v1beta1"
+	provcontrollers "github.com/rancher/rancher/pkg/generated/controllers/provisioning.cattle.io/v1"
+	"github.com/rancher/rancher/pkg/wrangler"
+)
+
+const (
+	removedAllPVCsAnnotationKey = "harvesterhci.io/removeAllPersistentVolumeClaims"
+)
+
+type handler struct {
+	capiClusters capicontrollers.ClusterCache
+	clusterCache provcontrollers.ClusterCache
+	secretCache  corecontrollers.SecretCache
+}
+
+func Register(ctx context.Context, clients *wrangler.Context) {
+	h := handler{
+		capiClusters: clients.CAPI.Cluster().Cache(),
+		clusterCache: clients.Provisioning.Cluster().Cache(),
+		secretCache:  clients.Core.Secret().Cache(),
+	}
+
+	clients.Dynamic.OnChange(ctx, "harvester-machine-provision-cleanup", validGVK, h.onChange)
```

Comment:
I am just wondering if it may be better to have a extra `OnRemove` handler on the provisioning cluster controller. This would ensure we have a extra finalizer for our cleanup, and we can check if the `.spec.machinePools[].machineConfigRef` contains a `HarvesterConfig` which would help us filter our Harvester specific clusters. 
Once we know the cluster is on Harvester, we can filter `rke-machines` based on cluster name and subsequently patch them.